server/src/main/java/org/apache/druid/segment/realtime/appenderator/BaseAppenderatorDriver.java

```diff
@@ -427,30 +428,36 @@ protected AppenderatorDriverAddResult append(
 
   /**
    * Returns a stream of {@link SegmentWithState} for the given sequenceNames.
+   *
+   * Note: it is not enough to synchronize on segments to access to this method,
+   * any operation on the resulting stream must also be synchronized on segments.
    */
+  @GuardedBy("segments")
   Stream<SegmentWithState> getSegmentWithStates(Collection<String> sequenceNames)
   {
-    synchronized (segments) {
-      return sequenceNames
-          .stream()
-          .map(segments::get)
-          .filter(Objects::nonNull)
-          .flatMap(segmentsForSequence -> segmentsForSequence.intervalToSegmentStates.values().stream())
-          .flatMap(segmentsOfInterval -> segmentsOfInterval.getAllSegments().stream());
-    }
+    return sequenceNames
```

Comment:
Since the only uses of these methods is just to get the segment identifier and collect it to a set 
```
...
        .map(SegmentWithState::getSegmentIdentifier)
        .collect(Collectors.toSet());
```
should these methods instead just return `Set<SegmentIdWithShardSpec>` instead of a stream so we can't make the error?

Reply:
we can definitely change the methods to do the full reduction. It seemed we had plans to use this for other things, but I can certainly simplify it

Reply:
@clintropolis I've update the code to return the fully materialized collections rather than streams.